Add support for named arguments in translations

Galarzaa90 opened the pull request:
This adds overloads to all translate functions to also accept a mapping of argument names to argument values:

**Example**:
```properties
command.apple.response=Hello my name is **{name}** and I have **{appleCount, number}** {appleCount, plural, =1{apple} other {apples}}.
```

```kt
translate("command.apple.response", mapOf("name" to "Allan", "appleCount" to 1))
> "Hello my name is **Allan** and I have **1** apple."
```

In some cases I extracted some code out current `translate` methods to avoid repeating it for the overload for named arguments.

Galarzaa90 committed:
Added support for named ICU message arguments

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/checks/types/CheckContext.kt:
```diff
@@ -153,11 +153,33 @@ public open class CheckContext<out T : Event>(public val event: T, public val lo
     /** Quick access to translate strings using this check context's [locale]. **/
     public fun translate(
         key: String,
-        bundle: String? = defaultBundle,
```
It seems this behaviour has not been preserved?

Galarzaa90 replied:
You're right, I have reverted this.

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/events/EventContext.kt:
```diff
@@ -78,4 +82,30 @@ public open class EventContext<T : Event>(
     ): String = translate(key, eventHandler.extension.bundle, replacements)
+
+    /**
+     * Given a translation key and possible replacements,return the translation for the given locale in the
```
Commas require spaces after.

Galarzaa90 replied:
Fixed all occurences of this, including the original ones.

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/events/EventContext.kt:
```diff
@@ -78,4 +82,30 @@ public open class EventContext<T : Event>(
+                return result
+            }
+        }
+        return null
```
You've removed a newline here.

Galarzaa90 replied:
Fixed this, but it didn't update in here for some reason, But if you check the commit, it is done.

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/events/EventHandler.kt:
```diff
@@ -278,4 +292,14 @@ public open class EventHandler<T : Event>(
     )
+
+    /**
+     * Given a translation key and possible replacements,return the translation for the given locale in the
```
Commas require spaces after.

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/i18n/ResourceBundleTranslations.kt:
```diff
@@ -120,7 +120,7 @@ public open class ResourceBundleTranslations(
     }
 
-    override fun translate(key: String, locale: Locale, bundleName: String?, replacements: Array<Any?>): String {
+    private fun getTranslatedString(key: String, locale: Locale, bundleName: String?): String {
```
This should not be `private` - mark it `protected` instead. Do not mark anything `private` on open classes.

Galarzaa90 replied:
Changed it to protected, and added kdoc string

gdude2002 commented on test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/I18nTestExtension.kt:
```diff
@@ -94,6 +95,23 @@ public class I18nTestExtension : Extension() {
+                respond {
+                    content = translate(
+                        "command.apple.response",
+                        mapOf(
```
This needs a newline before it.

Galarzaa90 replied:
Same here, it didn't update in here for some reason, But if you check the commit, it is done.

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/commands/CommandContext.kt:
```diff
@@ -114,4 +128,14 @@ public abstract class CommandContext(
     )
+
+    /**
+     * Given a translation key and possible replacements,return the translation for the given locale in the
```
Commas require spaces after.

gdude2002 reviewed (changes requested):
I'm seeing some breaking API changes and potential behavioural changes here - both of which would be unacceptable. Please ensure you test the i18n system in full with the bundled extensions.

Galarzaa90 committed:
Added tests for CheckContext and ValidationContext
- Added missing overloads for ValidationContext
- Fixed private method in open class
- Fixed some style issues

Galarzaa90 commented:
I covered all overloads of ValidationContext and CheckContext. Still need to add some tests for CommandContext.

Also, seems like I also missed adding overloads for ComponentContext.

Galarzaa90 committed:
Added translate overloads for ComponentContext

Galarzaa90 committed:
Added overload for Message.respondTranslated

Galarzaa90 commented:
Added missing overloads for ComponentContext now

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/checks/types/CheckContext.kt:
```diff
@@ -158,6 +158,28 @@ public open class CheckContext<out T : Event>(public val event: T, public val lo
+        replacements: Map<String, Any?>
+    ): String =
+        translations.translate(key, locale, bundleName = bundle, replacements = replacements)
+
```
It seems like you've created duplicate functions here - I feel like it'd make more sense to have functions that take a `bundle` argument that defaults to `defaultBundle` and not provide functions that take no bundle, right?

Galarzaa90 replied:
In the case of the first function in this block (key, replacements), this overload was already present in most of the other implementations ([example](https://github.com/Kord-Extensions/kord-extensions/pull/189/files/a305e9849b8ad40b3ed3c05d2caaea47f738edbd#diff-bb460ba4201be9a2852ab776870b77075a32408802712513759383971ae39456R81-R84)), but it was missing for CheckContext.

Without this function, you cannot call it like this:

```kt
// Doesn't compile, required String?, found Array... 
translate("check.positionalParameters", arrayOf(user.mention, user.id))

// This compiles
translate("check.positionalParameters", replacements = arrayOf(user.mention, user.id))
```

So the same is the case for the new functions with `Map`, without that overload with no bundles parameter, it forces you to use a named parameter, which is pretty annoying, and since this overload already existed for pretty much all other implementations, I also added it to these new translate methods.

gdude2002 replied:
:+1:

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/commands/converters/builders/ValidationContext.kt:
```diff
@@ -159,6 +159,28 @@ public class ValidationContext<out T>(public val value: T, public val context: C
+        replacements: Map<String, Any?>
+    ): String =
+        translations.translate(key, context.getLocale(), bundleName = bundle, replacements = replacements)
+
```
Again, it seems like you have duplicate functions here where the original file didn't have any duplicates.

Galarzaa90 replied:
This is another case where it was missing the overload with no bundle parameter for array replacements, as explained in my other comment.

gdude2002 commented on kord-extensions/src/main/kotlin/com/kotlindiscord/kord/extensions/i18n/ResourceBundleTranslations.kt:
```diff
@@ -59,7 +59,7 @@ public open class ResourceBundleTranslations(
     ): ResourceBundle = ResourceBundle.getBundle(bundle, locale, control)
 
```
As I believe this file has been modified since your PR was opened, can you double-check that the functionality provided by your changes matches the current logic?

Galarzaa90 replied:
Yes, I merged `develop` branch to my branch and tested again. Everything is working fine.

gdude2002 commented on test-bot/src/main/kotlin/com/kotlindiscord/kord/extensions/testbot/extensions/I18nTestExtension.kt:
```diff
@@ -94,6 +99,103 @@ public class I18nTestExtension : Extension() {
+        }
+    }
+
+    private suspend fun CheckContextWithCache<ChatInputCommandInteractionCreateEvent>.translatedChecks() {
```
If this is always intended to fail, please provide a comment to that effect, just so nobody is confused

Galarzaa90 replied:
Added the comments.

(I was thinking if I should even remove the response block and just leave an empty action block with the comment instead)

gdude2002 replied:
Nope, that's alright. Thanks!

gdude2002 reviewed (changes requested):
Sorry for taking so long to get to this - overall, this is looking pretty good, I don't see any glaring omissions.

Just a few small comments left, then - thanks for your patience!

Commit:
Merge remote-tracking branch 'origin/develop' into feat-translate-named-arguments

Galarzaa90 committed:
Add comments to commands that always expect to fail
